### PR TITLE
[DPO] tokenize and process DPO data via batches

### LINKED ISSUE #1310: Error when using multiprocessing in the DPO data preprocessing
dpo_trainer = DPOTrainer(
  File "/usr/local/lib/python3.10/dist-packages/trl/trainer/dpo_trainer.py", line 371, in __init__
    train_dataset = train_dataset.map(self.tokenize_row, num_proc=self.dataset_num_proc)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 592, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 557, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 3185, in map
    for rank, done, content in iflatmap_unordered(
  File "/usr/local/lib/python3.10/dist-packages/datasets/utils/py_utils.py", line 654, in iflatmap_unordered
    [async_result.get(timeout=0.05) for async_result in async_results]
  File "/usr/local/lib/python3.10/dist-packages/datasets/utils/py_utils.py", line 654, in <listcomp>
    [async_result.get(timeout=0.05) for async_result in async_results]
  File "/usr/local/lib/python3.10/dist-packages/multiprocess/pool.py", line 774, in get
    raise self._value
  File "/usr/local/lib/python3.10/dist-packages/multiprocess/pool.py", line 540, in _handle_tasks
    put(task)
  File "/usr/local/lib/python3.10/dist-packages/multiprocess/connection.py", line 209, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/local/lib/python3.10/dist-packages/multiprocess/reduction.py", line 54, in dumps
    cls(buf, protocol, *args, **kwds).dump(obj)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 418, in dump
    StockPickler.dump(self, obj)
  File "/usr/lib/python3.10/pickle.py", line 487, in dump
    self.save(obj)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/lib/python3.10/pickle.py", line 902, in save_tuple
    save(element)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/lib/python3.10/pickle.py", line 887, in save_tuple
    save(element)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 1212, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 1453, in save_instancemethod0
    pickler.save_reduce(MethodType, (obj.__func__, obj.__self__), obj=obj)
  File "/usr/lib/python3.10/pickle.py", line 692, in save_reduce
    save(args)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/lib/python3.10/pickle.py", line 887, in save_tuple
    save(element)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 1212, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 1212, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.10/pickle.py", line 713, in save_reduce
    self._batch_setitems(dictitems)
  File "/usr/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 1212, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.10/pickle.py", line 713, in save_reduce
    self._batch_setitems(dictitems)
  File "/usr/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.10/pickle.py", line 717, in save_reduce
    save(state)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 1212, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.10/pickle.py", line 713, in save_reduce
    self._batch_setitems(dictitems)
  File "/usr/lib/python3.10/pickle.py", line 1003, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 603, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.10/pickle.py", line 692, in save_reduce
    save(args)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/lib/python3.10/pickle.py", line 902, in save_tuple
    save(element)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 560, in save
    f(self, obj)  # Call unbound method with explicit self
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 1212, in save_module_dict
    StockPickler.save_dict(pickler, obj)
  File "/usr/lib/python3.10/pickle.py", line 972, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.10/pickle.py", line 998, in _batch_setitems
    save(v)
  File "/usr/local/lib/python3.10/dist-packages/dill/_dill.py", line 412, in save
    StockPickler.save(self, obj, save_persistent_id)
  File "/usr/lib/python3.10/pickle.py", line 578, in save
    rv = reduce(self.proto)
TypeError: cannot pickle 'torch._C._distributed_c10d.ProcessGroup' object  
  

 dpo_trainer = DPOTrainer(
  File "/usr/local/lib/python3.10/dist-packages/trl/trainer/dpo_trainer.py", line 369, in __init__
            dpo_trainer = DPOTrainer(dpo_trainer = DPOTrainer(dpo_trainer = DPOTrainer(

      File "/usr/local/lib/python3.10/dist-packages/trl/trainer/dpo_trainer.py", line 369, in __init__
  File "/usr/local/lib/python3.10/dist-packages/trl/trainer/dpo_trainer.py", line 369, in __init__
dpo_trainer = DPOTrainer(  File "/usr/local/lib/python3.10/dist-packages/trl/trainer/dpo_trainer.py", line 369, in __init__

  File "/usr/local/lib/python3.10/dist-packages/trl/trainer/dpo_trainer.py", line 369, in __init__
    with PartialState().local_main_process_first():
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
                with PartialState().local_main_process_first():with PartialState().local_main_process_first():
with PartialState().local_main_process_first():with PartialState().local_main_process_first():
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter_

### PR DESCRIPTION
**Pull request by kashif**
tokenize outside of the class and in batches 

fixes #1310 #1896 #1964

duplicate of PR #1896

### CODE REVIEW

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:71]: To be consistent with other implementations, we can allow `tokenizer` to be a processor here.

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:77]:
```suggestion
        if is_vision_model:
            processor = tokenizer
```

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:655]:
```suggestion
                fn_kwargs={"tokenizer": self.processor if self.is_vision_model else self.tokenizer, "is_vision_model": self.is_vision_model},
```

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:92]: Why don't you have images here?

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:83]: We use batch here, so we don't want to "unbatch"

```suggestion
```

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:92]: It concatenates in the batch dim, while we want to concat each elements in the batch 

```python
prompt = ["My name is", "I live in"]
chosen = ["Quentin", "Lyon"]
print(prompt + chosen)
# ["My name is", "I live in", "Quentin", "Lyon"]
```

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:660]: I would keep `writer_batch_size=10`. Otherwise, it uses a batch size of 1000, which, when using images, can cause more than 2GB to be written at once, and raise an exception.
Furthermore, in my experience, lowering this value doesn't make the code any slower.

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:669]: same here: https://github.com/huggingface/trl/pull/1914/files#r1713735488

**Review by lewtun — COMMENTED** [review]
Thanks for the refactor @kashif ! Overall LGTM barring the points @qgallouedec raised (which I agree with). 

Since this is a change with potentially large impact, can you do a comparison of the loss / reward curves of the `DPOTrainer` on `main` vs this branch to make sure we have parity and are not introducing any subtle regressions?

**Comment by kashif** [comment]
yeah will make sure to test

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:195]: How do you handle the attention mask here?

File: trl/trainer/dpo_trainer.py
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:227]:
```suggestion
        if "pixel_attention_mask" in full_tokenized:
            result["pixel_attention_mask"] = full_tokenized["pixel_attention_mask"]
```

**Comment by kashif** [comment]
ran with:

```
 python examples/scripts/dpo.py \                                                                                              
    --dataset_name=trl-internal-testing/hh-rlhf-helpful-base-trl-style \
    --model_name_or_path=gpt2 \
    --per_device_train_batch_size 4 \
    --learning_rate 1e-3 \
    --gradient_accumulation_steps 1 \
    --logging_steps 10 \
    --eval_steps 500 \
    --output_dir="dpo_anthropic_hh" \
    --warmup_steps 150 \
    --report_to wandb \
    --bf16 \
    --logging_first_step \
    --no_remove_unused_columns \
    --dataset_num_proc 8
```
see run [here](https://wandb.ai/krasul/huggingface/runs/ip1mqj02).

File: trl/trainer/dpo_trainer.py
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:72]: Can you add a unit test for this please?

File: trl/trainer/dpo_trainer.py
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:101]: Can you add a unit test for this please?

File: trl/trainer/dpo_trainer.py
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:194]: Can you add a unit test for this please?

**Review by lewtun — COMMENTED** [review]
Thanks for adding the Wandb logs! Could you also run the same thing on `main` so we can overlay the loss / reward curves?

I also realised with this refactor we can unit test the processing functions - would you mind adding some tests? Check out the KTO tests as an example

**Comment by qgallouedec** [comment]
LGTM. I would recommend to check there is no regression before merging though

**Review by qgallouedec — APPROVED** [review]

**Comment by lewtun** [comment]
I ran the `dpo.py` example with DDP on both the `main` branch and this PR. I find there are some slight discrepancies between the two, which you can find in these runs:

* Without refactor: https://wandb.ai/huggingface/huggingface/runs/6nqn730n
* With refactor: https://wandb.ai/huggingface/huggingface/runs/6nqn730n?nw=nwuserlewtun

Screenshot 2024-08-19 at 15 15 07
Screenshot 2024-08-19 at 15 14 06

Here's the command I ran:

```shell
accelerate launch --config_file=examples/accelerate_configs/multi_gpu.yaml examples/scripts/dpo.py \
    --dataset_name=trl-internal-testing/hh-rlhf-helpful-base-trl-style \
    --model_name_or_path=gpt2 \
    --per_device_train_batch_size 4 \
    --learning_rate 1e-3 \
    --gradient_accumulation_steps 1 \
    --logging_steps 10 \
    --eval_steps 500 \
    --output_dir="dpo_anthropic_hh" \
    --warmup_steps 150 \
    --report_to wandb \
    --bf16 \
    --logging_first_step \
    --no_remove_unused_columns
```

@kashif it would be good to find the root cause of these differences - one way to do this would be to have a unit test that uses the old tokenization method and compares against the new. Once they match, you can remove the old methods

**Review by lewtun — APPROVED** [review]
LGTM once the WandB logs showing no regressions are linked.

**Comment by kashif** [comment]
run with refactoring: https://wandb.ai/krasul/huggingface/runs/484hg85j
run without refactoring: https://wandb.ai/krasul/huggingface/runs/a8uozmc2

**Comment by lewtun** [comment]
I took a look at your two runs @kashif and it looks good! We have some minor differences in the margins / loss, but I think that can be attributed to noise

Screenshot 2024-08-28 at 15 57 45
Screenshot 2024-08-28 at 15 57 38

**Pull request merged**
